wip: feature/wrapper

SirMallard opened the pull request:
This is a huge rewrite of large swathes of Iris. The main focus of this was to unify much of the behaviour, reduce any unncessary behaviour and change all the APIs to functions, rather than tables, which should be much easier to use.

⚠️ This change is monumental and a breaking change, and will require changing all existing Iris implementation in order to upgrade.

### This has involved:
- removing large parts of the types system, designed to make it easier to add more to the library by focusing less on static types, and alllowing the Luau type solver to generate itself, which is much easier for development
- renamed many existing behaviours, such as unifying the defintions of opened and closed for windows, trees and similar widgets.
- moved to the new Content type for all images
- removed some deprecated or unused APIs
- decomposed the library into individual files to reduce inter-dependency. This fixes all the modules returning a single function requiring Iris as an argument in order to add more behaviour
- hid away more non-user behaviour
- added all widget types to the documentation
- introduced new bit flags, which are for representing multiple boolean arguments. This is most noticable for Windows, which reduces 8 arguments into 1. These flags are named within a table in Iris, and can just be combined using an addition operation
- renamed a bunch of states to better refelct or shorten the names, ie. isChecked -> checked. Additionally, the .value property has been changed to ._value
- any properties or functions which are not meant to be public for users and now prefixed with an underscore to signify they are meant for general use
- the arguments provided to a widget include both regular arguments and states, which requires a number of regular arguments in order to separate them out
- changed Disabled to _paused internally, which should make it clearer what the behaviour is. Will probably add a function to change the value

There are still a lot of changes to be made, mostly for checking that the documentation is all correct and that the behaviour is still correct. Hopefully this will make Iris easier to use for both developers and myself. I'm going to leave this request out for a while, in order to gather feedback.

The release of this will be v3.0.0 because of the many breaking changes.

SirMallard committed:
Added half of widgets.

SirMallard committed:
Finished wrapper.

SirMallard committed:
Changed to GuiBase2d as parent.

SirMallard committed:
Merge branch 'main' into feature/wrapper

SirMallard committed:
Rewrote parts of the library.

SirMallard committed:
Updated widget.

SirMallard committed:
Added API.

SirMallard committed:
Moved over to lib.

SirMallard committed:
Updated some of the docs.

SirMallard committed:
Added more types.

SirMallard committed:
Updated more docs.

SirMallard committed:
Finished docs move.

SirMallard committed:
Cleanup.

SirMallard committed:
Updated stylua.

SirMallard committed:
Reworked most of the demo window.

SirMallard committed:
Added new CheckboxFlags.

SirMallard committed:
Updated checks workflow versions.

SirMallard committed:
Fixed config names.

SirMallard committed:
Changed events.

SirMallard committed:
Updated documentation.

SirMallard committed:
Changed file names.

SirMallard committed:
Fixed API transition issues.

SirMallard committed:
Formatting and docs.

SirMallard committed:
Merge branch 'main' into feature/wrapper

SirMallard committed:
Merge branch 'main' into feature/wrapper

SirMallard committed:
Changed require settings.

SirMallard committed:
Merge branch 'main' into feature/wrapper

SirMallard committed:
Updated starter docs.

SirMallard committed:
Merge branch 'main' into feature/wrapper

SirMallard commented:
This feature introduces a lot of major, breaking changes to the library and changes the API a lot. Lots of these changes are for consistency between widgets.

## Events
Widget events are used to monitor a change in a widget. This is useful for buttons, where a user would want to listen to the button being clicked. Along the way, we have introduced many common events for widgets, such listening to when a Tree is collapsed or uncollapsed. With the modern day API, we can detect this two ways:
```lua
local tree = Iris.Tree("Collapsing Tree")

-- using the widget event
if tree.collapsed() then
  -- internally checks tree.lastCollapsedTick == Iris._cycleTick
end
-- using the state event
if tree.state.collapsed:changed() and not tree.state.collapsed then
-- internally checks tree.state.collapsed.lastChangeTick == Iris._cycleTick
end
```

In this example here, we can see that our second example is very verbose, although it achieves the same behaviour, just using different methods. We are effectively duplicating code here for no reason. At a minimum, we should change our event to use the state `:changed()` event to remove the need for an extra tick counter.

But this introduces our second question: what events should we include? For a tree, our `hovered`, `collapsed` and `uncollapsed` events are for detecting a change, but for a RadioButton we include an `active` event, which determines if this radio button is selected. I would not classify this as an 'event', even though it is a necessary addition. Additionally, we include a state for a PlotLines, which is readonly for the currently hovered value. Should we instead put this as event?

Therefore, how many events should we include, Should we include the minimum, non-state evenfs only, such as `hovered` and `clicked`? Or should we also include wrappers around some state events? Should we also include state checks, such as when a window is opened and closed **and** to check if the window is in the open or close state? At this point, should events include general widget queries as well?

## States
States are very important for widget functionality and many widgets share common state behaviours. For example, both Trees and Windows can be collapsed. Menus and Combos can be opened and closed. But is there any difference between this behaviour?

Not only are some state names quite long, but they also overlap with events. Some widgets use shared states, which introduces other naming difficulties. They also seem inconsistent, such as a datatype input using a `value` state but the InputText using a `text` state.

I propose a new naming convention, which simplifies state names to be more consistent and obvious:
```markdown
### Window
= size: State<Vector2>
= position: State<Vector2>
= isUncollapsed: State<boolean> -> open
= isOpened: State<boolean>      -> shown
= scrollDistance: State<number> -> scroll

### Menu
= isOpened: State<boolean>

### MenuToggle
= isChecked: State<boolean>     -> value

### InputText
= text: State<string>           -> value

### Checkbox
= isChecked: State<boolean>     -> value

### RadioButton
= index: State<any>

### Tree, CollapsingHeader
= isUncollapsed: State<boolean> -> open

### TabBar
= index: State<number>

### Tab
= isOpened: State<boolean>      -> open

### Input<T>, Drag<T>, Slider<T> { number, Vector2, Vector3, UDim, UDim2, Rect }
= number: State<T>         -> value
= editingText: State<boolean>   - [EVENT]

### InputColor3
= color: State<Color3>
= editingText: State<boolean>   - [EVENT]

### InputColor4
= color: State<Color3>
= transparency: State<number>
= editingText: State<boolean>   - [EVENT]

### Selectable
= index: State<any>

### Combo, ComboArray, ComboEnum
= index: State<any>,
= isOpened: State<boolean>      -> open

### InputCombo
= number: State<number>         -> value
= editingText: State<boolean>   - [EVENT]
= enumItem: EnumItem ?

### ProgressBar
= progress: State<number>

### PlotLines, PlotHistorgram
= values: State<{number}>
= hovered: State<{number}>      - [EVENT]

### Table
= widths: State<{number}>
```

The most notable changes are:
- moving all the Input, Drag and Slider widgets to use a `value` state, rather than a `number` state
- renaming the `isUncollapsed` states into `open` states
- renaming the `isChecked` states into `value` states

I would like to know what people think of these changes.

JakeyWasTaken commented:
I agree with renaming the `isChecked`, `number`, `text` fields to something more unified like `value`. But I think changing these three on window is overkill

```
= isUncollapsed: State<boolean> -> open
= isOpened: State<boolean>      -> shown
= scrollDistance: State<number> -> scroll
```

both `open` and `shown` are somewhat synonymous. As for `scrollDistance`, I've never used it but again I don't see the reason to drop the 8 letters and kill verbosity when simply reading the type name atm will explain all of them super clearly.

Maybe if they were shared across a bunch of different widgets but 🤷‍♂️ 

Something more akin to 

```
= isUncollapsed: State<boolean> -> collapsed
= isOpened: State<boolean>      -> open
```

I would agree with

SirMallard commented:
> I agree with renaming the `isChecked`, `number`, `text` fields to something more unified like `value`. But I think changing these three on window is overkill
> 
> ```
> = isUncollapsed: State<boolean> -> open
> = isOpened: State<boolean>      -> shown
> = scrollDistance: State<number> -> scroll
> ```
> 
> both `open` and `shown` are somewhat synonymous. As for `scrollDistance`, I've never used it but again I don't see the reason to drop the 8 letters and kill verbosity when simply reading the type name atm will explain all of them super clearly.
> 
> Maybe if they were shared across a bunch of different widgets but 🤷‍♂️
> 
> Something more akin to
> 
> ```
> = isUncollapsed: State<boolean> -> collapsed
> = isOpened: State<boolean>      -> open
> ```
> 
> I would agree with

Thanks for that opinion. I agree that `open` and `shown` are very similar which would be a pain to work with. For something like a Tree, I don't know whether `open` would be a better word. Especially with other similar items, such as Combos or Menus, where they are `open` or `collapsed`?

TaylorsRus commented:
i concur, that simply swapping to using the affirmative form, `isUncollapsed` for example is a bit round-about; let the user flip the form in their code
 
I'd still want to keep the `is` convention, so `isCollapsed`, and `isVisible` would be my reccomendations. When we lose the `is`, the line between booleans and other state representations can blur. 

`visible` to me works, where `open` doesn't, because a collapsed menu could also be referred to as closed, however a collapsed menu could not be referred to as invisible; We also mirror common roblox terminology for GuiObject.Visible in that regard, which can make the intent even clearer.

I think `shown` works just aswell as `visible`, in the case that `collapsed` is used to refer to the other, similar state

SirMallard committed:
Changing state names.

SirMallard committed:
Beginning work on test cases.

Commit:
Merge branch 'main' into feature/wrapper

Commit:
Updated more properties.

Commit:
Added Lute and API changes.

Added Lute to support generating documenation and types. Moved over the 'text' widget argument to be a 'label' instead, to better reflect its purpose.

Commit:
Updated and fixed type warnings.
